DepositRegistrar.sol

```diff
@@ -0,0 +1,306 @@
+        entry name = hashedNames[bid.hashedName()];
+        
+        if (bid.creationDate() > name.auctionDate - revealPeriod
+            || now > name.auctionDate ) {
+            // bid is invalid, burn 99.9%
+            bid.closeDeed(999);
+            
+        } else if ( _value < minBid ) {
+            // bid is invalid but not punishable, burn 0.1%
+            bid.closeDeed(1);
+            
+        } else if (bid.value() > name.highestBid) {
+            // new winner
+            // cancel the other bid, burn 0.1%
+            Deed previousWinner = Deed(name.deed);
+            previousWinner.closeDeed(1);
+            
+            // set new winner
+            name.value = name.highestBid;
+            name.highestBid = bid.value();
+            name.deed = sealedBids[seal];
+        
+        } else if (bid.value() > name.value) {
+            // not winner, but affects second place
+            name.value = bid.value();
+            bid.closeDeed(1);
+            
+        } else {
+            // bid doesn't affect auction
+            bid.closeDeed(1);
+        }
+    }
+    
+    function finalizeAuction(bytes32 nameHash) noEther {
+        entry name = hashedNames[nameHash];
+        if (now < name.auctionDate || name.status != Mode.Auction) throw;
+        
+        // set the name
+        name.status = Mode.Owned;
+        name.firstRegistered = now;
+        name.lastRenewed = now;
+
+        //set the averages
+        uint period = (now - lastSinceNewName) * M;
+        averagePeriod = (9999 * averagePeriod + period) / 10000;
+        uint n = averagePeriod < (60 seconds * M) ? 7 days * M / averagePeriod : 10000;
+        averagePrice = ((n-1)*averagePrice + name.value ) / n;
+        name.averagePrice = averagePrice;
+        lastSinceNewName = now;
+        
+        Deed deedContract = Deed(name.deed);
+        
+        deedContract.setBalance(name.value > minBid ? name.value : minBid);
+    }
+    
+    
+    /*
+    ## Renewals can be done at any moment by renewing the deposit
+
+    In order to renew a name you need to calculate how much have average prices of 
+    names changed since you last bought it. If the average price names have increased, 
+    then you need to increase your deposit. If median prices have fallen down, then 
+    you have the right to withdraw part of the difference (1/2 after a year, 3/4 
+    after 2 years, 7/8 after 3 years etc). 
+    */ 
+    function updatedValue(bytes32 nameHash) constant returns (uint updatedPrice) {
+        entry name = hashedNames[nameHash];
+        return name.value * averagePrice / name.averagePrice;
+    }
+    
+    function renewDeed(bytes32 nameHash) {
+        entry name = hashedNames[nameHash];
+        if (now < name.firstRegistered + renewalPeriod/2 ) throw;
+        Deed deedContract = Deed(name.deed);
+        uint difference = 0;
+        if (msg.sender != deedContract.owner() && name.status != Mode.Owned) throw;
```

Comment:
With &&, this will only throw if you're both not the owner and it's not owned, though. If the name is owned by someone else, this will evaluate to (false && false), and fail to throw.